Add support for inheritance from class with different suffix in modular

yonigozlan opened the pull request:
# What does this PR do?
Add support for inheritance from class with different suffix in modular, fixes https://github.com/huggingface/transformers/issues/33900#issuecomment-2391616600

Useful when we want to create a model for a certain task, extending another model used for a different task, as is the case for `ColPaliForRetrieval` and `PaliGemmaForConditionalGeneration` in https://github.com/huggingface/transformers/pull/33736

Not sure if that's in the spirit of modular or if it's the best way to do this though!

<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons---sometimes notifications get lost.
-->

<!-- Remove if not applicable -->

Fixes #33900

## Who can review?

cc @ArthurZucker , @tonywu71 
<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

Models:

- text models: @ArthurZucker
- vision models: @amyeroberts, @qubvel
- speech models: @ylacombe, @eustlb
- graph models: @clefourrier

Library:

- flax: @sanchit-gandhi
- generate: @zucchini-nlp (visual-language models) or @gante (all others)
- pipelines: @Rocketknight1
- tensorflow: @gante and @Rocketknight1
- tokenizers: @ArthurZucker
- trainer: @muellerzr and @SunMarc
- chat templates: @Rocketknight1

Integrations:

- deepspeed: HF Trainer/Accelerate: @muellerzr
- ray/raytune: @richardliaw, @amogkam
- Big Model Inference: @SunMarc
- quantization (bitsandbytes, autogpt): @SunMarc

Documentation: @stevhliu

HF projects:

- accelerate: [different repo](https://github.com/huggingface/accelerate)
- datasets: [different repo](https://github.com/huggingface/datasets)
- diffusers: [different repo](https://github.com/huggingface/diffusers)
- rust tokenizers: [different repo](https://github.com/huggingface/tokenizers)

Maintained examples (not research project or legacy):

- Flax: @sanchit-gandhi
- PyTorch: See Models above and tag the person corresponding to the modality of the example.
- TensorFlow: @Rocketknight1

 -->

yonigozlan committed:
add support for different suffix in modular

ArthurZucker reviewed (commented):
Overall sounds good, I applied similar changes here: https://github.com/huggingface/transformers/pull/33962/commits/a2a6a9ba1f7395f0428d80d1ec405c286881e099 which you have probably seen! 

Can you try to import them (ex: raise the error earlier, use camel case replace, no old_class_name and new_class_name for simplification etc) 

and most importantly, work with a dummy example this way we can see exactly what you are enabling. 

Hardest cases are when you inherit from CLIPMLP or CLIPTextAttention for MyModelAttention

yonigozlan committed:
add dummy example, pull new changes for modular

yonigozlan committed:
nide lines order change

yonigozlan commented:
No I hadn't seen those changes sorry about that, I have added them now.
Also added a dummy example very similar to what we're trying to do with ColPali!

ArthurZucker reviewed (approved):
Yup nice thanks for updating! 🤗 LGTM
